source/common/upstream/upstream_impl.cc

```diff
@@ -629,11 +629,19 @@ StaticClusterImpl::StaticClusterImpl(const envoy::api::v2::Cluster& cluster,
     : ClusterImplBase(cluster, cm.bindConfig(), runtime, stats, ssl_context_manager, added_via_api),
       initial_hosts_(new HostVector()) {
 
-  for (const auto& host : cluster.hosts()) {
-    initial_hosts_->emplace_back(HostSharedPtr{new HostImpl(
-        info_, "", resolveProtoAddress(host), envoy::api::v2::core::Metadata::default_instance(), 1,
-        envoy::api::v2::core::Locality().default_instance(),
-        envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance())});
+  const auto& load_assignment = cluster.has_load_assignment()
+                                    ? cluster.load_assignment()
+                                    : Config::Utility::translateClusterHosts(cluster.hosts());
+
+  const auto& locality_lb_endpoints = load_assignment.endpoints();
+  for (const auto& locality_lb_endpoint : locality_lb_endpoints) {
+    for (const auto& lb_endpoint : locality_lb_endpoint.lb_endpoints()) {
+      const auto& host = lb_endpoint.endpoint().address();
+      initial_hosts_->emplace_back(HostSharedPtr{new HostImpl(
+          info_, "", resolveProtoAddress(host), envoy::api::v2::core::Metadata::default_instance(),
+          1, envoy::api::v2::core::Locality().default_instance(),
```

Comment:
Why ignore locality here? This might be surprising. We should probably document which combination of cluster type and features work.